ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -0,0 +1,165 @@
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.of(StoreErrorCodes.ID_Deleted,
+        StoreErrorCodes.ID_Deleted_Permanently, StoreErrorCodes.ID_Not_Found);
+    checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if permanent blob is permanent in backups as well.
+   * This method does not expect any errors when retrieving the blob from store.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyTtlUpdate(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.TTL_UPDATE);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
+    checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if an un-deleted blob has been un-deleted from backups as well.
+   * The backup could be lagging behind servers and the blob could be temp or perm state.
+   * This is acceptable as well since un-delete takes the blob back to one of these states.
+   * And if we recover from a lagging backup in case of a disaster, we are consistent from a user's point of view.
+   * TODO: If we the blob was permanent before un-delete, then it should be permanent after un-delete.
+   * TODO: Same for a temporary blob. Distinguish between these two cases.
+   * @param onPremBlob the {@link MessageInfo} that will be transformed into an undelete
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyUndelete(MessageInfo onPremBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.PUT,
+        MessageInfoType.TTL_UPDATE, MessageInfoType.UNDELETE);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.of(StoreErrorCodes.ID_Undeleted);
+    checkBackup(onPremBlob, remoteReplicaInfo, acceptableBackupBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if a temporary blob is present in the backup.
+   * The replica could be lagging behind the backup and the blob could be permanent or even deleted from the backup.
+   * The replica should catch up with its peers. Nonetheless, this check will catch such replicas as well.
+   * @param validMessageDetectionInputStream Stream of valid blob IDs
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyPut(MessageSievingInputStream validMessageDetectionInputStream,
+      RemoteReplicaInfo remoteReplicaInfo) {
+    EnumSet<MessageInfoType> acceptableBackupBlobStates = EnumSet.of(MessageInfoType.PUT);
```

Comment:
Yes, you are right. I just read the replicaThread code and all the cases and your comment makes more sense now. If we are here, then that means the blob is not present in the local store.

Reply:
overrode fixMissingStoreKeys instead of applyPut. Please take a look and lmk what you think.

Reply:
Can we follow the current code and make sure we don't miss any code?
For example, 
fixMissingStoreKeys will call createGetRequest. 
createGetRequest will check
"if (exchangeMetadataResponse.serverErrorCode == ServerErrorCode.No_Error)" condition etc.

Reply:
yeah i followed the code. Since i override it, this version will not send any getRequests and will not receive any response. It will simply check the backup and log the missing key.

Reply:
ok, i see what u mean, will fix it